compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2622,9 +2762,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     return expected;
                 }
                 InheritedRefMatchRule::EatInner => {
-                    if let ty::Ref(_, _, r_mutbl) = *expected.kind()
-                        && pat_mutbl <= r_mutbl
-                    {
+                    let expected_ref_or_pinned_ref = || {
+                        if self.tcx.features().pin_ergonomics()
+                            && let Some(ty::Ref(_, _, r_mutbl)) =
+                                expected.pinned_ty().map(|ty| *ty.kind())
+                            && pat_mutbl <= r_mutbl
+                        {
+                            return Some((Pinnedness::Pinned, r_mutbl));
+                        }
+                        if let ty::Ref(_, _, r_mutbl) = *expected.kind()
+                            && pat_mutbl <= r_mutbl
+                        {
+                            return Some((Pinnedness::Not, r_mutbl));
+                        }
+                        None
+                    };
+                    if let Some((_, r_mutbl)) = expected_ref_or_pinned_ref() {
```

Comment:
I'm confused, this should be allowing `let &mut foo: &pin mut T`, but you have a test that shows that this gives a type mismatch error. Do you understand why? If not that, what's the purpose of this change here?

Reply:
Oh it's that this applies only to `InheritedRefMatchRule::EatInner` which is behavior gated under `ref_pat_eat_one_layer_2024_structural`. You have to support `EatBoth` for stable rust (you can probably ignore the other two variants for now, they're for feature gates). Anyway I think it's better not to allow `&mut x` to eat a `&pin mut T` for now, let's leave this to a future PR.